packages/sdk-react/src/components/approvalRequests/ApprovalRequests.tsx

```diff
@@ -0,0 +1,116 @@
+import { useIsActionAllowed } from '@/core/queries/usePermissions';
+import { AccessRestriction } from '@/ui/accessRestriction';
+import { ActionEnum } from '@/utils/types';
+import { t } from '@lingui/macro';
+import { useLingui } from '@lingui/react';
+import { CircularProgress } from '@mui/material';
+import { useQueryClient } from '@tanstack/react-query';
+
+import { ApprovalRequestsTable } from './ApprovalRequestsTable';
+
+export const ApprovalRequests = () => (
+  <MoniteScopedProviders>
+    <ApprovalRequestsBase />
+  </MoniteScopedProviders>
+);
+
+const ApprovalRequestsBase = () => {
+  const { i18n } = useLingui();
+  const { api } = useMoniteContext();
+  const queryClient = useQueryClient();
+
+  const { data: user } = useEntityUserByAuthToken();
+
+  const { data: isReadAllowed, isLoading: isReadAllowedLoading } =
+    useIsActionAllowed({
+      method: 'approval_request',
+      action: ActionEnum.READ,
+      entityUserId: user?.id,
+    });
+
+  const approveRequestMutation =
+    api.approvalRequests.postApprovalRequestsIdApprove.useMutation();
+  const rejectRequestMutation =
+    api.approvalRequests.postApprovalRequestsIdReject.useMutation();
+
+  const onApprove = (approvalRequestId: string) => {
+    approveRequestMutation.mutate(
+      {
+        body: undefined,
+        path: {
+          approval_request_id: approvalRequestId,
+        },
+      },
+      {
+        onSuccess: () => {
+          toast.success(t(i18n)`Request was approved`);
+
+          return api.approvalRequests.getApprovalRequests.invalidateQueries(
+            queryClient
+          );
+        },
+      }
+    );
+  };
+
+  const onReject = (approvalRequestId: string) => {
+    rejectRequestMutation.mutate(
+      {
+        body: undefined,
+        path: {
+          approval_request_id: approvalRequestId,
+        },
+      },
+      {
+        onSuccess: () => {
+          toast.success(t(i18n)`Request was rejected`);
+
+          return api.approvalRequests.getApprovalRequests.invalidateQueries(
+            queryClient
+          );
+        },
+      }
+    );
+  };
+
+  return (
+    <>
+      <PageHeader
+        title={
+          <>
+            {t(i18n)`Approval Requests`}
+            {isReadAllowedLoading && (
+              <CircularProgress size="0.7em" color="secondary" sx={{ ml: 1 }} />
+            )}
+          </>
+        }
+      />
+      {!isReadAllowedLoading && !isReadAllowed && <AccessRestriction />}
+      {isReadAllowed && (
+        <ApprovalRequestsTable
+          onRowActionClick={({ id, action }) => {
+            switch (action) {
+              case 'approve':
+                onApprove(id);
+                break;
+              case 'reject':
+                onReject(id);
+                break;
+              default:
+                break;
+            }
+          }}
+          isApprovePending={approveRequestMutation.isPending}
+          isRejectPending={rejectRequestMutation.isPending}
```

Comment:
Instead of adding extra properties to indicate loading status, what if we return a promise and manage the loading state within the component? For example:

```tsx
<ApprovalRequestsTable
  onRowActionClick={({ id, action }) => {
    switch (action) {
      case 'approve':
        return approveRequestMutation.mutateAsync(id);
      case 'reject':
        return rejectRequestMutation.mutateAsync(id);
      default:
        throw new Error('Unhandled approval action');
    }
  }}
/>
```

We can then handle the loading state like this:

```ts
setIsSomeActionLoading(true);
onRowActionClick(...).finally(() => setIsSomeActionLoading(false));
```

Reply:
I rewrote actions in this way